components/ledger/ledger-verification/src/main/kotlin/net/corda/ledger/verification/sandbox/impl/VerificationSandboxServiceImpl.kt

```diff
@@ -28,51 +29,40 @@ import org.slf4j.LoggerFactory
 /**
  * This is a sandbox service that is internal to this component.
  *
- * It gets/creates a sandbox with a per-sandbox:
- *
- *   * serializer
- *   * entity manager factory
+ * It gets/creates a Verification sandbox with a per-sandbox serializers.
  *
  */
-@Suppress("LongParameterList")
 @RequireSandboxAMQP
 @RequireSandboxJSON
 @Component(service = [ VerificationSandboxService::class ])
 class VerificationSandboxServiceImpl @Activate constructor(
     @Reference
     private val sandboxService: SandboxGroupContextComponent,
+    @Reference
+    private val cpkReadService: CpkReadService
 ) : VerificationSandboxService {
     companion object {
         private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
     }
 
     override fun get(holdingIdentity: HoldingIdentity, cpks: List<CordaPackageSummary>): SandboxGroupContext {
-
-        val cpkFileChecksums = cpks.mapTo(mutableSetOf()) { SecureHash.parse(it.fileChecksum) }
-        if (!sandboxService.hasCpks(cpkFileChecksums)) {
-            throw CpkNotAvailableException(
-                "CPKs for Verification Sandbox for $holdingIdentity are not available (yet): $cpks"
-            )
-        }
-
-        val sandbox = sandboxService.getOrCreate(getVirtualNodeContext(holdingIdentity, cpkFileChecksums)) { _, ctx ->
+        checkCpks(holdingIdentity, cpks)
+        return sandboxService.getOrCreate(getVirtualNodeContext(holdingIdentity, cpks)) { _, ctx ->
             initializeSandbox(holdingIdentity, ctx)
         }
+    }
 
-        // Only contract CPKs are allowed
-        val nonContractCpks = sandbox.sandboxGroup
-            .metadata
-            .values
-            .filterNot { it.isContractCpk() }
-
-        if (nonContractCpks.isNotEmpty()) {
-            val cpkIds = nonContractCpks.map { it.cpkId }
-            throw NotAllowedCpkException(
-                "CPKs for Verification Sandbox for $holdingIdentity are not allowed: $cpkIds"
+    private fun checkCpks(holdingIdentity: HoldingIdentity, cpks: List<CordaPackageSummary>) {
+        cpks.forEach {
+            val cpk = cpkReadService.get(it.fileChecksum.toSecureHash()) ?: throw CpkNotAvailableException(
+                "CPK for Verification Sandbox for $holdingIdentity is not available (yet): $it"
             )
+            if (!cpk.metadata.isContractCpk()) {
+                throw NotAllowedCpkException(
+                    "CPK for Verification Sandbox for $holdingIdentity is not allowed: $it"
```

Comment:
I don't like this error message, it's  not very clear. How about something like this:
```suggestion
                    "This CPK is not allowed for use in the Verification Sandbox for $holdingIdentity - only contract CPKs are allowed: $it"
```